consensus/src/chained_bft/event_processor.rs

```diff
@@ -63,6 +65,79 @@ mod event_processor_test;
 #[path = "event_processor_fuzzing.rs"]
 pub mod event_processor_fuzzing;
 
+/// During the event of the node can't recover from local data, StartupSyncProcessor is responsible
+/// for processing the events carrying sync info and use the info to retrieve blocks from peers
+#[allow(dead_code)]
+pub struct StartupSyncProcessor<T> {
+    network: NetworkSender,
+    storage: Arc<dyn PersistentLivenessStorage<T>>,
+    ledger_recovery_data: LedgerRecoveryData<T>,
+}
+
+#[allow(dead_code)]
+impl<T: Payload> StartupSyncProcessor<T> {
+    pub fn new(
+        network: NetworkSender,
+        storage: Arc<dyn PersistentLivenessStorage<T>>,
+        ledger_recovery_data: LedgerRecoveryData<T>,
+    ) -> Self {
+        StartupSyncProcessor {
+            network,
+            storage,
+            ledger_recovery_data,
+        }
+    }
+
+    pub async fn process_proposal_msg(
+        &mut self,
+        proposal_msg: ProposalMsg<T>,
+    ) -> Result<RecoveryData<T>> {
+        let author = proposal_msg.proposer();
+        let sync_info = proposal_msg.sync_info();
+        self.sync_up(&sync_info, author).await
+    }
+
+    pub async fn process_vote(&mut self, vote_msg: VoteMsg) -> Result<RecoveryData<T>> {
+        let author = vote_msg.vote().author();
+        let sync_info = vote_msg.sync_info();
+        self.sync_up(&sync_info, author).await
+    }
+
+    pub async fn process_sync_info_msg(
+        &mut self,
+        sync_info: SyncInfo,
+        peer: Author,
+    ) -> Result<RecoveryData<T>> {
+        debug!("Startup Sync received a sync info msg: {}", sync_info);
+        self.sync_up(&sync_info, peer).await
+    }
+
+    async fn sync_up(&mut self, sync_info: &SyncInfo, peer: Author) -> Result<RecoveryData<T>> {
+        ensure!(
+            sync_info.highest_round() > self.ledger_recovery_data.commit_round(),
+            "Received sync info has lower round number than committed block"
+        );
+        ensure!(
+            sync_info.epoch() == self.ledger_recovery_data.epoch(),
+            "Received sync info is in different epoch than committed block"
+        );
+        let mut retriever = BlockRetriever::new(
+            self.network.clone(),
+            // Give a timeout of 1 hour to make sure there's enough time for trying to fetch from all peers
+            Instant::now() + Duration::new(3600, 0),
```

Comment:
we should never wait for such long to get 3 blocks, which means a malicious node could block this node for an hour

Reply:
Oh right, the timeout here is per attempt and not the overall timeout across the attempts.

Reply:
@zekun000 Sorry I missed that, what would be a reasonable time per attempt?